src/Forms/GridField/GridFieldSudoMode.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+namespace SilverStripe\Forms\GridField;
+
+use SilverStripe\Forms\FormField;
+use SilverStripe\Forms\SudoModePasswordField;
+use SilverStripe\View\ArrayData;
+use SilverStripe\View\SSViewer;
+
+/**
+ * GridFieldSudoMode warning that the managed data is protected by SudoMode which will
+ * need to be activated before the data can be edited.
+ *
+ * This is particularly important for GridFields that contain no data
+ */
+class GridFieldSudoMode extends AbstractGridFieldComponent implements GridField_HTMLProvider
+{
+    /**
+     *
+     * @param GridField $gridField
+     * @return array
+     */
+    public function getHTMLFragments($gridField)
+    {
+        $form = $gridField->getForm();
+        // Do not render this component if the entire form is protected by sudo mode.
+        // If the form is protected then there will already be a SudoModePasswordField as the first form field
+        if ($form->getFormRequiresSudoMode()) {
+            return [];
+        }
+        $field = SudoModePasswordField::create();
+        $field->setCollapsed(true);
+        // Form needs to be set otherwise the field will not render correctly
+        $field->setForm($form);
+        // Should use silverstripe/admin/themes/cms-forms/templates/SilverStripe/Forms/FormField_holder.ss
+        // The lowercase "e" in "extraClass" is necessary to match the template
+        // It will be used by SudoModePasswordFieldEntwine.js to initialise the
+        $holderTemplateData = ArrayData::create([
+            'extraClass' => 'SudoModePasswordField SudoModePasswordField--collapsed',
```

Comment:
This class should be added just by calling `setCollapsed(true)` on the field - otherwise that setter is only doing half of its job.

Either amend the template to add the class:
`<% if $Collapsed %>SudoModePasswordField--collapsed<% end_if %>`
or update the setter to set and remove the extra class
```php
if ($collapsed) {
    $this->addExtraClass('SudoModePasswordField--collapsed');
} else {
    $this->removeExtraClass('SudoModePasswordField--collapsed');
}
```
I'm not sure what happens if you call `addExtraClass()` with the same class multiple times so probably just updating the template is better.

We should be able to just call `$field->FieldHolder()` here to get the full and correct rendered result, with no need for the new `GridFieldSudoMode.ss` template.

Reply:
NOTE: You may want to look at the feedback in admin first.

Reply:
Have updated gone with putting the logic on SudoModePasswordField::extraClass()